libraries/plugins/transaction_status/transaction_status_plugin.cpp

```diff
@@ -48,43 +81,152 @@ void transaction_status_impl::on_post_apply_block( const block_notification& not
+         return block->transactions.front().id();
+   }
+
+   return {};
+}
+
+/**
+ * Retrieve the latest transaction_id in a given range.
+ *
+ * \param[in] first_block_num The first block in the provided range
+ * \param[in] last_block_num The last block in the provided range
+ * \return The transaction_id of the latest transaction that *should* be in our transaction status index
+ */
+fc::optional< transaction_id_type > transaction_status_impl::get_latest_transaction_in_range( const uint32_t first_block_num, const uint32_t last_block_num )
+{
+   for ( uint32_t block_num = last_block_num; block_num >= first_block_num; block_num-- )
+   {
+      const auto block = _db.fetch_block_by_number( block_num );
+
+      if ( block.valid() && block->transactions.size() > 0 )
+         return block->transactions.back().id();
+   }
+
+   return {};
+}
+
+/**
+ * Determine if the plugin state is valid.
+ *
+ * We determine validity by checking to see if we are aware of the first and last transactions
+ * in the range we are supposed to be tracking.
+ *
+ * \return True if the transaction state is considered valid, otherwise false
+ */
+bool transaction_status_impl::state_is_valid()
+{
+   const auto head_block_num = _db.head_block_num();
+   uint32_t earliest_tracked_block_num = get_earliest_tracked_block_num();
+
+   const auto earliest_tx = get_earliest_transaction_in_range( earliest_tracked_block_num, head_block_num );
+   const auto latest_tx = get_latest_transaction_in_range( earliest_tracked_block_num, head_block_num );
+
+   bool lower_bound_is_valid = true;
+   bool upper_bound_is_valid = true;
+
+   if ( earliest_tx.valid() )
+      lower_bound_is_valid = _db.find< transaction_status_object, by_trx_id >( *earliest_tx ) != nullptr;
+
+   if ( latest_tx.valid() )
+      upper_bound_is_valid = _db.find< transaction_status_object, by_trx_id >( *latest_tx ) != nullptr;
+   
+   return lower_bound_is_valid && upper_bound_is_valid ? true : false;
+}
+
+/**
+ * Rebuild the necessary transaction status index objects.
+ *
+ * If the transaction status plugin is enabled at some arbitrary point in time, it may not have
+ * the necessary data to track transaction unless the state is rebuilt. We must clear out the
+ * existing objects and recreate the objects as if it would have during run-time.
+ */
+void transaction_status_impl::rebuild_state()
+{
+   ilog( "Rebuilding transaction status state" );
+
+   // Clear out the transaction status index
+   const auto& tx_status_idx = _db.get_index< transaction_status_index >().indices().get< by_trx_id >();
+   for ( auto& e : tx_status_idx )
+      _db.remove( e );
+
+   // Re-build the index from scratch
+   const auto head_block_num = _db.head_block_num();
+   uint32_t earliest_tracked_block_num = get_earliest_tracked_block_num();
+
+   for ( uint32_t block_num = earliest_tracked_block_num; block_num <= head_block_num; block_num++ )
+   {
+      const auto block = _db.fetch_block_by_number( block_num );
+
```

Comment:
Add `FC_ASSERT( block.valid(), "Could not read block ${n}", ("n", block_num) );`